robogen/src/main.rs

```diff
@@ -0,0 +1,84 @@
+use std::io::Write;
+
+
+#[derive(Debug, Copy, Clone)]
+struct PixelState {
+    hvis: bool,
+    hsync: bool,
+    vvis: bool,
+    vsync: bool,
+    frame_vis: bool,
+    end: bool
+}
+
+impl Default for PixelState {
+    fn default() -> Self {
+        PixelState { hvis: false, hsync: true, vvis: false, vsync: true, frame_vis: false, end: false }
+    }
+}
+
+impl From<PixelState> for u8 {
+    fn from(p: PixelState) -> u8 {
+        p.hvis as u8 | (p.hsync as u8) << 1 | (p.vvis as u8) << 2 | (p.vsync as u8) << 3 | (p.frame_vis as u8) << 4 | (p.end as u8) << 5
+    }
+}
+
+const WIDTH: usize = 800 / 16;
+const HEIGHT: usize = 525;
+
+
+fn main() {
+    let mut timing = [PixelState::default(); WIDTH * HEIGHT + 1];
+
+    for line in 0..HEIGHT {
+        for pixel in 0..WIDTH {
+            let i = line * WIDTH + pixel;
+            let mut state = PixelState::default();
+            let mut repchar = '_';
+            if pixel < 40 {
+                state.hvis = true;
+            }
+            if pixel >= 41 && pixel < 47 {
+                // active low
+                state.hsync = false;
+            }
+            
+            if line >= 40 && line < 440 {
+                state.vvis = true;
+            }
+            if line >= 490 && line < 492 {
+                // active low
+                state.vsync = false;
+            }
+            state.frame_vis = state.hvis && state.vvis;
+            if state.frame_vis {
+                repchar = '.';
+            }
+            if !state.hsync || !state.vsync {
+                repchar = '^';
+            }
+            eprint!("{}", repchar);
+            timing[i] = state;
+        }
+        eprintln!();
+    }
+    let endstate = PixelState { end: true, ..Default::default() };
+    timing[WIDTH * HEIGHT] = endstate;
```

Comment:
I think this should be:

```suggestion
    timing[WIDTH * HEIGHT - 1].end = true;
``` 

The way the [schematic](https://github.com/rustbox/cuddly-robot/pull/3#issuecomment-1292936157) has it wired, we've got to go through the latch before the counters will see the reset signal, so I think this will be a `CLK_sync` pulse too late.

Reply:
did we decide if wiring end directly into reset if we made it active low would work? (I think it would)

Reply:
Yeah, that's right, as long as we never touch /OE or /CE on the EEPROM (which would disconnect the reset line). From our discussion (edited):

> The latch grabs the last address' data, so the output will see the data from address 0 typically about 4ns after the clock reads 1, because the typical CLK -> Q update for our 74HC163s is ~24ns, but whereas the typical timing for the 74HC574s is ~28ns). The reset is asynchronous, so wherever we put the bit gets dropped: from the time the counters assert the address, the EEPROM emits the low signal in ~150ns, the counters reset, and the EEPROM reads from address 0 in another 150ns, well before the next rising edge 650ns later.